### PR TITLE
Import Flow-3.7 to external Flow release

### PR DESCRIPTION
**Pull request by kongw95**
*Release the internal Flow-3.7*

**Main changes**
1.  Graceful shutdown. When a deployment or whatsoever comes to a host where workflow/activity worker is running, and the worker successfully polls a decision/activity task but cannot finish the task because the host is bounced, the workflow execution will get stuck until reaching the decision/activity task start_to_close_timeout. To address this issue, we add graceful shutdown into SpringWorkflowWorker and SpringActivityWorker, which will stop polling and wait for the current task to complete before initiating shutdown.
2.  Separate threads for polling and executing decider logic. The trigger for this change is to avoid hitting the hard limit (1000) on concurrent polling per task-list. Customers can keep the total polling thread count under 1000 for their workflow worker fleet, while having more than 1000 threads for executing decider logic. 

**Test**
I've verified that the changes do not break any of our Flow integration tests.

### CODE REVIEW

File: src/main/java/com/amazonaws/services/simpleworkflow/flow/WorkflowContext.java [hunk 1]
```diff
@@ -60,15 +60,25 @@
...
      * <pre>
+     <<<<<<< HEAD
      * if (workflowContext.isImplementationVersion(COMPONENT_NAME, 5)) {
      *     // New code path
      * }
      * else if (workflowContext.isImplementationVersion(COMPONENT_NAME, 4)) {
      *     // Old version 4 code path
      * }
      * else if (workflowContext.isImplementationVersion(COMPONENT_NAME, 3)) {
+     =======
```
Comment (jorgesiq) [inline src/main/java/com/amazonaws/services/simpleworkflow/flow/WorkflowContext.java:73]: Probably during conflict resolution you forgot to fix this. Please fix it.

File: src/main/java/com/amazonaws/services/simpleworkflow/flow/WorkflowContext.java [hunk 2]
```diff
@@ -60,15 +60,25 @@
      * This method is created to enable workflow updates without changing their
      * type version. The code changes should follow the following pattern:
      * <p>
-     * 
+     *
      * <pre>
+     <<<<<<< HEAD
```
Comment (jorgesiq) [inline src/main/java/com/amazonaws/services/simpleworkflow/flow/WorkflowContext.java:65]: Probably during conflict resolution you forgot to fix this. Please fix it.

File: src/main/java/com/amazonaws/services/simpleworkflow/flow/WorkflowContext.java [hunk 3]
```diff
@@ -93,18 +103,26 @@
      * executed the second call to isVersion returns <code>false</code> even if
      * it is executed for the first time.
      * <p>
-     * 
+     *
      * <pre>
+     <<<<<<< HEAD
```
Comment (jorgesiq) [inline src/main/java/com/amazonaws/services/simpleworkflow/flow/WorkflowContext.java:108]: Probably during conflict resolution you forgot to fix this. Please fix it.

File: src/main/java/com/amazonaws/services/simpleworkflow/flow/WorkflowContext.java [hunk 4]
```diff
@@ -93,18 +103,26 @@
...
      * if (workflowContext.isImplementationVersion(&quot;comp1&quot;, 1)) {
+     =======
```
Comment (jorgesiq) [inline src/main/java/com/amazonaws/services/simpleworkflow/flow/WorkflowContext.java:110]: Probably during conflict resolution you forgot to fix this. Please fix it.

File: src/main/java/com/amazonaws/services/simpleworkflow/flow/WorkflowContext.java [hunk 5]
```diff
@@ -93,18 +103,26 @@
...
+     * if (workflowContext.isVersion(&quot;comp1&quot;, 1)) {
+     >>>>>>> d7ef09c... save for test use.
```
Comment (jorgesiq) [inline src/main/java/com/amazonaws/services/simpleworkflow/flow/WorkflowContext.java:112]: Probably during conflict resolution you forgot to fix this. Please fix it.

File: src/main/java/com/amazonaws/services/simpleworkflow/flow/WorkflowContext.java [hunk 6]
```diff
@@ -93,18 +103,26 @@
...
      * }
      * else {
      *     // Old code path 1
      * }
-     * 
+     *
      * // Location of the workflow execution when upgrade was deployed.
-     * 
+     *
+     <<<<<<< HEAD
```
Comment (jorgesiq) [inline src/main/java/com/amazonaws/services/simpleworkflow/flow/WorkflowContext.java:121]: Probably during conflict resolution you forgot to fix this. Please fix it.

File: src/main/java/com/amazonaws/services/simpleworkflow/flow/WorkflowContext.java [hunk 7]
```diff
@@ -93,18 +103,26 @@
...
      * if (workflowContext.isImplementationVersion(&quot;comp1&quot;, 1)) {
+     =======
+     * if (workflowContext.isVersion(&quot;comp1&quot;, 1)) {
+     >>>>>>> d7ef09c... save for test use.
```
Comment (jorgesiq) [inline src/main/java/com/amazonaws/services/simpleworkflow/flow/WorkflowContext.java:125]: Probably during conflict resolution you forgot to fix this. Please fix it.

File: src/main/java/com/amazonaws/services/simpleworkflow/flow/WorkflowContext.java (hunk 1 above)
Comment (kongw95) [reply src/main/java/com/amazonaws/services/simpleworkflow/flow/WorkflowContext.java]: Fixed.

File: src/main/java/com/amazonaws/services/simpleworkflow/flow/WorkflowContext.java (hunk 2 above)
Comment (kongw95) [reply src/main/java/com/amazonaws/services/simpleworkflow/flow/WorkflowContext.java]: Fixed.

File: src/main/java/com/amazonaws/services/simpleworkflow/flow/WorkflowContext.java (hunk 3 above)
Comment (kongw95) [reply src/main/java/com/amazonaws/services/simpleworkflow/flow/WorkflowContext.java]: Fixed.

File: src/main/java/com/amazonaws/services/simpleworkflow/flow/WorkflowContext.java (hunk 4 above)
Comment (kongw95) [reply src/main/java/com/amazonaws/services/simpleworkflow/flow/WorkflowContext.java]: Fixed.

File: src/main/java/com/amazonaws/services/simpleworkflow/flow/WorkflowContext.java (hunk 5 above)
Comment (kongw95) [reply src/main/java/com/amazonaws/services/simpleworkflow/flow/WorkflowContext.java]: Fixed.

File: src/main/java/com/amazonaws/services/simpleworkflow/flow/WorkflowContext.java (hunk 6 above)
Comment (kongw95) [reply src/main/java/com/amazonaws/services/simpleworkflow/flow/WorkflowContext.java]: Fixed.

File: src/main/java/com/amazonaws/services/simpleworkflow/flow/WorkflowContext.java (hunk 7 above)
Comment (kongw95) [reply src/main/java/com/amazonaws/services/simpleworkflow/flow/WorkflowContext.java]: Fixed.

**Pull request merged**